File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/PICSTokensCallback.java
package in.dragonbra.javasteam.steam.handlers.steamapps.callback;

import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserverAppinfo.CMsgClientPICSAccessTokenResponse;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.JobID;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * This callback is fired when the PICS returns access tokens for a list of appids and packageids
 */
public class PICSTokensCallback extends CallbackMsg {

    private final List<Integer> packageTokensDenied;

    private final List<Integer> appTokensDenied;

    private final Map<Integer, Long> packageTokens;

    private final Map<Integer, Long> appTokens;

    public PICSTokensCallback(JobID jobID, CMsgClientPICSAccessTokenResponse.Builder msg) {
        setJobID(jobID);

        packageTokensDenied = Collections.unmodifiableList(msg.getPackageDeniedTokensList());
        appTokensDenied = Collections.unmodifiableList(msg.getAppDeniedTokensList());

        packageTokens = new HashMap<>();
        appTokens = new HashMap<>();

        for (CMsgClientPICSAccessTokenResponse.PackageToken packageToken : msg.getPackageAccessTokensList()) {
            packageTokens.put(packageToken.getPackageid(), packageToken.getAccessToken());
        }

        for (CMsgClientPICSAccessTokenResponse.AppToken appToken : msg.getAppAccessTokensList()) {
            appTokens.put(appToken.getAppid(), appToken.getAccessToken());
        }
    }

    /**
     * @return a list of denied package tokens.
     */
    public List<Integer> getPackageTokensDenied() {
        return packageTokensDenied;
    }

    /**
     * @return a list of denied app tokens.
     */
    public List<Integer> getAppTokensDenied() {
        return appTokensDenied;
    }

    /**
     * @return a map containing requested package tokens.
     */
    public Map<Integer, Long> getPackageTokens() {
        return packageTokens;
    }

    /**
     * @return a map containing requested package tokens.
     */
    public Map<Integer, Long> getAppTokens() {
        return appTokens;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/PICSTokensCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamapps.callback

import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserverAppinfo.CMsgClientPICSAccessTokenResponse
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg

/**
 * This callback is fired when the PICS returns access tokens for a list of appids and packageids
 */
class PICSTokensCallback(packetMsg: IPacketMsg) : CallbackMsg() {

    /**
     * Gets a list of denied package tokens.
     */
    val packageTokensDenied: List<Int>

    /**
     * Gets a list of denied app tokens.
     */
    val appTokensDenied: List<Int>

    /**
     * Gets a map containing requested package tokens.
     */
    val packageTokens: Map<Int, Long>

    /**
     * Gets a map containing requested package tokens.
     */
    val appTokens: Map<Int, Long>

    init {
        val tokensResponse = ClientMsgProtobuf<CMsgClientPICSAccessTokenResponse.Builder>(
            CMsgClientPICSAccessTokenResponse::class.java,
            packetMsg
        )
        val msg = tokensResponse.body

        jobID = tokensResponse.targetJobID

        packageTokensDenied = msg.packageDeniedTokensList
        appTokensDenied = msg.appDeniedTokensList

        packageTokens = msg.packageAccessTokensList.associate { it.packageid to it.accessToken }
        appTokens = msg.appAccessTokensList.associate { it.appid to it.accessToken }
    }
}
